python/pyarrow/builder.pxi

```diff
@@ -42,7 +42,7 @@ cdef class StringBuilder(_Weakrefable):
         value : string/bytes or np.nan/None
             The value to append to the string array builder.
         """
-        if value is None or value is np.nan:
+        if value is None or ('numpy' in sys.modules and value is np.nan):
```

Comment:
Not directly related to this PR (and I assume this code is not used much), but it's also a bit strange that this checks explicitly for np.nan in addition to None (it could also check for `float("nan")`, for example), and additionally checking for NaN with `is np.nan` is also wrong because that doesn't always give true .. (eg `np.array([np.nan])[0] is np.nan` is False)